semgrep-core/src/extracting/Extract.ml

```diff
@@ -39,82 +39,269 @@ let ( let* ) = Option.bind
+  end_pos : int;
+}
+
+let count_lines_and_trailing =
+  String.fold_left
+    (fun (n, c) b ->
+      match b with
+      | '\n' -> (n + 1, 0)
+      | _ -> (n, c + 1))
+    (0, 0)
+
+let offsets_of_mval extract_mvalue =
+  Metavariable.mvalue_to_any extract_mvalue
+  |> Visitor_AST.range_of_any_opt
+  |> Option.map (fun (start_loc, end_loc) ->
+         let end_len = String.length end_loc.Parse_info.str in
+         {
+           start_pos = start_loc.charpos;
+           (* subtract 1 because lines are 1-indexed, so the
+               offset is one less than the current line *)
+           start_line = start_loc.Parse_info.line - 1;
+           start_col = start_loc.column;
+           end_pos = end_loc.charpos + end_len;
+         })
+
+(*****************************************************************************)
+(* Error reporting *)
+(*****************************************************************************)
+
+let report_unbound_mvar ruleid mvar m =
+  let { Range.start; end_ } = Pattern_match.range m in
+  logger#warning
+    "The extract metavariable for rule %s (%s) wasn't bound in a match; \
+     skipping extraction for this match [match was at bytes %d-%d]"
+    ruleid mvar start end_
+
+let report_no_source_range erule =
+  logger#error
+    "In rule %s the extract metavariable (%s) did not have a corresponding \
+     source range"
+    (fst erule.Rule.id)
+    (let (`Extract { Rule.extract; _ }) = erule.mode in
+     extract)
+
+(*****************************************************************************)
+(* Result mapping helpers *)
+(*****************************************************************************)
+
+let map_loc pos line col file (loc : Parse_info.token_location) =
+  (* this _shouldn't_ be a fake location *)
+  {
+    loc with
+    charpos = loc.charpos + pos;
+    line = loc.line + line;
+    column = (if loc.line = 1 then loc.column + col else loc.column);
+    file;
+  }
+
+let map_res map_loc tmpfile file
+    (mr : Report.partial_profiling Report.match_result) =
+  {
+    Report.matches =
+      Common.map
+        (fun (m : Pattern_match.t) ->
+          { m with file; range_loc = Common2.pair map_loc m.range_loc })
+        mr.matches;
+    errors =
+      Common.map
+        (fun (e : Semgrep_error_code.error) -> { e with loc = map_loc e.loc })
+        mr.errors;
+    skipped_targets =
+      Common.map
+        (fun (st : Output_from_core_t.skipped_target) ->
+          { st with path = (if st.path = tmpfile then file else st.path) })
+        mr.skipped_targets;
+    profiling = { mr.profiling with Report.file };
+  }
+
 (*****************************************************************************)
 (* Main logic *)
 (*****************************************************************************)
 
-(** This is the main function which performs extraction of the matches
-   generated by extract mode rules.
+let extract_collect_and_comb erule_table xtarget rule_ids matches =
```

Comment:
This function is pretty brutal, but you've done a nice job making it elegant. I might add a comment above each `|> ...` summarizing what that step is doing (at least the ones that are more than like 2 lines).